### ISSUE TITLE
where to get the vocab file for the pretrained librispeech language model

### ISSUE THREAD

**Issue by Slyne** [rwth-i6/returnn-experiments#12]
Hi, 
I was wondering where I could find the vocab file for the pretrained language model since I only found the language model for librispeech.

data_files = {
    "train": "/work/asr3/irie/data/librispeech/lm_bpe/librispeech-lm-norm.bpe.txt.gz",
    "cv": "/work/asr3/irie/data/librispeech/lm_bpe/dev.clean.other.bpe.txt.gz",
    "test": "/work/asr3/irie/data/librispeech/lm_bpe/test.clean.other.txt.gz"}
vocab_file = "/work/asr3/irie/data/librispeech/lm_bpe/trans.bpe.vocab.lm.txt"

For the train, cv, test files, I could use apply_bpe.py to generate the according datasets but if I want to use the pretrained model, I should use the same vocabulary file as the pretrained model. Not sure if I miss something.

**Comment by albertz** [comment]
The vocabulary created by the apply_bpe script is the same as for the pretrained model.

**Comment by albertz** [comment]
I'm but sure what you mean by librispeech-lm-norm.txt.
You just use the vocabulary (vocab_file) which you created via full-setup-attention.
You don't need anything else.
You can use the configs [here](https://github.com/rwth-i6/returnn-experiments/tree/master/2018-asr-attention/librispeech/lm) to do recognition with your attention model together with a LM (e.g. the pretrained model).

**Comment by Slyne** [comment]
I mean in the pretrained model config files you give, there are four files
which be given.
train: which is the training language model corpus, should be trained on
librispecch-lm-norm.txt
the vocab file should be a txt file, and the format should be:
word  index
which is different from what I get from the full-setup-attention, which
looks like this:
{
'<s>': 0,
'</s>': 0,
'<unk>': 1,
'THE': 2,
'AND': 3,
'OF': 4,
'TO': 5,
'A': 6,
'IN': 7,
.....
}


Attach part of the pretrained model configuration file:
confile:
data_files = {
"train":
"/work/asr3/irie/data/librispeech/lm_bpe/librispeech-lm-norm.bpe.txt.gz",
"cv": "/work/asr3/irie/data/librispeech/lm_bpe/dev.clean.other.bpe.txt.gz",
"test": "/work/asr3/irie/data/librispeech/lm_bpe/test.clean.other.txt.gz"}
vocab_file =
"/work/asr3/irie/data/librispeech/lm_bpe/trans.bpe.vocab.lm.txt"}

Albert Zeyer <notifications@github.com> 於 2018年12月29日 週六 上午12:08寫道：

**Comment by Slyne** [comment]
I need to train the language model. Furthermore, the vocab file has two
symbols with the same index which will cause error when using the config
file.

Albert Zeyer <notifications@github.com> 於 2019年1月2日 週三 上午6:27寫道：

**Comment by Slyne** [comment]
And the config file seems to be a training configuration. You can notice that the task="train" in the configuration file.

**Comment by albertz** [comment]
Ah, I thought that you want to use the pretrained model.
If you want to train it yourself, then yes, you need that config (btw, yes, there is `task='train'`, but still in many cases, the config file is used for both training + inference; any option there can be overwritten via command line).
I just checked and you are right, the vocab used for the `LmDataset` has a different format. But I think it should be straight forward to convert from one to the other. Or to add support for the other format. Actually I wonder why we have that at all. I uploaded that vocab file [here](http://www-i6.informatik.rwth-aachen.de/~zeyer/models/librispeech/lm/bpe-10k/trans.bpe.vocab.lm.txt).
The train files (data_files in config) are generated somehow from the [LibriSpeech LM training data](http://www.openslr.org/11/). We might have done some post processing / normalization. @kazuki-irie can give you more details on that. Actually, maybe he can just upload his normalization scripts also [here](https://github.com/rwth-i6/returnn-experiments/tree/master/2018-asr-attention/librispeech/lm).

**Comment by Slyne** [comment]
Thanks!
That's clear now.

Albert Zeyer <notifications@github.com> 於 2019年1月2日 週三 下午4:02寫道：

**Comment by albertz** [comment]
No problem. I extended the Readme a bit ([here](https://github.com/rwth-i6/returnn-experiments/tree/master/2018-asr-attention/librispeech/lm)).
Maybe @kazuki-irie can later add some of the post processing scripts.
Closing this now.
